.emacs.d/config.org
#+TITLE: Emacs configuration file
#+AUTHOR: spook
#+BABEL: :cache yes
#+PROPERTY: header-args :tangle yes

* TODO
  - [ ] choose consistent format to package/PL setup
    - probably means difference between use-package'ing everything and making
      config blocks with more explicit annotations :(. I don't think
      use-package lends itself to literate programming as well, but it arguably
      leads to better package management and configuration overall.

* my-init
  configurations for my emacs. these configs are opinionated and by no
  guarantee "good." that said, i decided to return to vanilla emacs because it
  ensures i can debug issues and problems when they occur, something i lost in
  the midst of spacemac’s complexity. ergo, the return of my =.emacs= file.

  this emacs borrows code directly from and/or took inspiration from the following users:
  - [[https://github.com/larstvei/dot-emacs][Lars]]
  - [[https://writequit.org/org/][Lee]]
  - [[https://github.com/purcell/emacs.d][Purcell]]


  i have learned an incredible amount about what nice/clean elisp looks like from these user configs
  and do a horrible job of reflecting that below.

** overview
*** philosophy
    the main of philosophy of this config is as follows, in order of
    importance:

    1. readability
       - without readability all is lost in an emacs config. everything will
         inevitably soon crumble after losing the ability to easily understand
         what is happening in the configuration.

    2. usability
       - it should work, damn it. nothing weird, and err towards simplicity
         over complexity

    3. modularity
       - avoid global configurations where uncalled for, e.g. less =(global-<MODE>)=

       - configuration functions are good, let’s use more of them!

*** how this works
    unlike Lar's configuration, my =init.el= will always be called. i like the
    idea of =init.el= always remaining untouched and pure.

    in =init.el=, i set my package repositories and then bootstrap
    =use-package= and =org=, these are the two pillars of my config and i
    want to guarantee they are working and up to date.

    from there, the init file loads this file (=my-init.org=), extracts all
    code blocks by "tangling" it, and produces the =.el= equivalent used by
    emacs.

**** use-package notes
***** =:init= vs =:config=

      - =:init= is executed upon the eval of a =(use-package ...)= declaration
        containing them, i.e. immediately when *this* file is evaluated by
        emacs, *always*.
        - not affected by whether the package is lazy loaded or not.

      - =:config= is executed *when* and only *if* the given package in a
        use-package declaration is evaluated. This is because of the lazy
        nature of package configurations.

      - Thus, you will typically want to use =:config= for settings specific
        the mode being loaded, e.g. =(setq some-mode-color ’red)=

*** TODO Decide whether I will byte compile my emacs config
    we want our changes in any =.org= config file to be reflected back into our actual =.el= files. The
    hook below achieves this by re-compiling the byte file of any config =.org= file edited.

    #+BEGIN_SRC emacs-lisp :tangle no
    (defun tangle-init ()
      "If the current buffer is 'config.org' the code-blocks are
    tangled, and the tangled file is compiled."
      (when (equal (buffer-file-name)
                   (expand-file-name (concat user-emacs-directory "config.org")))
        ;; Avoid running hooks when tangling.
        (let ((prog-mode-hook nil))
          (org-babel-tangle)
          (byte-compile-file (concat user-emacs-directory "config.el")))))

    (add-hook 'after-save-hook 'tangle-init)
    #+END_SRC

**** =use-package=
     always use =use-package= for loading up packages due to its lazy nature. this means emacs doesn't
     load something until it actually needs it, saving us a lot of time and some RAM by cutting boot
     times and pre-loading things when we might never use them.

** the basics
*** TODO packages
    one single action on this section:
    - [ ] remove all the shit i just don't need/use.
      + defer to use-package declarations to do the work for me, if anything.

    the spice of emacs begins with packages:

    #+BEGIN_SRC emacs-lisp
      (let* ((package--builtins nil)
             (packages
              '(
                ;; utilities
                ivy swiper counsel fill-column-indicator volatile-highlights
                golden-ratio undo-tree ace-window paredit projectile counsel-projectile diminish rainbow-delimiters
                smartparens ggtags expand-region keychain-environment workgroups2 multiple-cursors
                dimmer smooth-scrolling exec-path-from-shell unfill xterm-color
                pandoc-mode lorem-ipsum quelpa-use-package alarm-clock

                ;; keybindings
                general which-key


                ;; auto-complete
                company flycheck flycheck-tip yasnippet auto-yasnippet flycheck-irony company-c-headers
                elisp-slime-nav company-math company-auctex company-anaconda
                slime-company company-ycmd flycheck-rust company-web
                company-ghci flycheck-haskell company-statistics yasnippet-snippets


                ;; lsp stuff
                ;; removed dap-mode for debugging over modeline and tooltip
                ;; same goes for lsp-treemacs
                lsp-mode lsp-ui lsp-ivy

                ;; git
                magit diff-hl

                ;; markdown
                markdown-mode toml-mode yaml-mode

                ;; lua
                lua-mode

                ;; c/cpp
                clang-format cmake-mode

                ;; rust
                cargo racer rust-mode

                ;; lisp
                slime racket-mode

                ;; clojure
                cider cider-eval-sexp-fu clj-refactor clojure-mode clojure-snippets clojure-mode-extra-font-locking

                ;; haskell
                haskell-mode haskell-snippets hindent hlint-refactor

                ;; python
                anaconda-mode pyenv-mode hy-mode pytest py-yapf ;; python-pytest

                ;; TODO: actually set these up
                js-doc js2-mode js2-refactor rjsx-mode json-mode json-snatcher web-beautify coffee-mode vue-mode
                tide nodejs-repl

                ;; TODO: actually set these up to be used.
                ;; webprogramming
                haml-mode sass-mode scss-mode slim-mode web-mode less-css-mode pug-mode emmet-mode

                ;; elm
                elm-mode

                ;; tex
                auctex auctex-latexmk typo cdlatex

                ;; org
                org-bullets org-pomodoro org-download htmlize org-journal secretaria org-download toc-org

                ;; themes
                challenger-deep-theme seti-theme gandalf-theme cyberpunk-theme dark-mint-theme
                kaolin-themes minimal-theme smart-mode-line smart-mode-line-powerline-theme

                ;; other
                writeroom-mode minimap gnuplot csv-mode fireplace anki-editor emojify define-word

                )))
        (let ((packages (remove-if 'package-installed-p packages)))
          (when packages
            ;; Install uninstalled packages
            (package-refresh-contents)
            (mapc 'package-install packages))))
    #+END_SRC
**** local packages
     setup the directory used for custom packages, i.e. stuff that isn't typically up on melpa but i
     might find on github, the emacs wiki, or i might write myself
     #+BEGIN_SRC emacs-lisp
       (let ((default-directory (concat (expand-file-name (concat user-emacs-directory "site-lisp/")))))
         (normal-top-level-add-subdirs-to-load-path))
     #+END_SRC
**** quelpa/quelpa-use-package
     These allow me to install emacs packages from non-traditional sources, i.e. git repos and
     whatnot. It's a lot more powerful than that, but tbh i'll probably only ever use it for git
     repos.
     #+BEGIN_SRC emacs-lisp
       (use-package quelpa)
         ;; this lets me use quelpa as if its use-package
       (use-package quelpa-use-package)
     #+END_SRC
*** general settings

    things that don't seem to go anywhere because of their generality or edge-case nature:
**** Fonts

     someone put together an emacs package mode that seems to make fira code actually play nicely
     and without requiring tons of messing around (i've lost literal days to this lol).

     #+BEGIN_SRC emacs-lisp
       (use-package fira-code-mode
         :hook prog-mode
         :config
         (setq fira-code-mode-disabled-ligatures '("x")))
     #+end_src

     #+BEGIN_SRC emacs-lisp
       ;; using gccemacs built with gtk, doesn't give me issues apparently so i'm re-using it
       (setq x-gtk-use-system-tooltips t)

     #+END_SRC

**** system defaults

    #+BEGIN_SRC emacs-lisp
      ;; need this for things like pyenv to work with emacs daemon.
      ;; this is a recent development and apparently has something to do with the fact
      ;; i am using the daemon as a systemd unit. that shouldn't be a problem. to checkout later.
      (exec-path-from-shell-initialize)

      ;; no splash screen, sorry Stallman
      (setq inhibit-splash-screen t)

      ;; disables annoying noise, "bell", triggered with actions such as hitting
      ;; the end of a buffer.
      (setq visible-bell t)

      ;; ensure things like gpg and ssh agent are handled automatically
      (keychain-refresh-environment)


      ;; don't use the tool or scroll bar, thx
      (tool-bar-mode -1)
      (scroll-bar-mode -1)

      ;; sometimes I do like the menu bar, but not currently
      (menu-bar-mode -1)

      ;; scrolling speed is bad, make it nicer
      ;; NOTE: It appears smooth scrolling has a very perverse effect on
      ;; erc mode by hogging up the cpu by a significant %. Disabling for now.
      ;; (smooth-scrolling-mode 1)
      (setq mouse-wheel-progressive-speed nil)

      ;; UTF-8 is our friend in a world of shitty programming standards
      (set-language-environment "UTF-8")
      (prefer-coding-system 'utf-8)
      (set-default-coding-systems 'utf-8)
      (set-terminal-coding-system 'utf-8)
      (set-keyboard-coding-system 'utf-8)
      ;;(setq default-buffer-file-coding-system 'utf-8)

      ;; speaking of text, unify emacs clipboard with linux
      (setq x-select-enable-clipboard t)

      ;; global hl line mode
      (global-hl-line-mode t)

      ;; can't get flycheck mode to enable itself otherwise
      (global-flycheck-mode)

      ;; make our mode line prettier
      (use-package smart-mode-line
        :config
        (setq sml/no-confirm-load-theme t
              ;; As it stands, I actually like the dark theme
              ;; in conjunction with minimal-light more than sml's light theme
              sml/theme 'dark)
        (sml/setup))


      ;; make sure it attempts utf-8 first when pasting text into emacs
      (setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

      ;; always tell me the column and row of where i am please, i am already lost
      ;; enough as is.
      (line-number-mode 1)
      (column-number-mode 1)

      ;; do you seriously think i'm good at typing
      (setq read-file-name-completion-ignore-case t)

      ;; life is too short for fully authenticating bad decisions
      (defalias 'yes-or-no-p 'y-or-n-p)

      ;; fix bad escape sequence weirdness
      (setq system-uses-terminfo nil)

      ;; stronk encryption good
      ;;------------------------
      ;; TODO figure out how to make this compatible with erc-tls because
      ;; 4096 encryption makes any connection impossible because freenode uses
      ;; sub-4096 encryption
      ;; (setq gnutls-min-prime-bits 4096)

      ;; don't let me be moronic and kill emacs w/o warning. GUI only.
      (when (window-system)
        (setq confirm-kill-emacs 'yes-or-no-p))

      ;; move through lines based on visual appearance rather than literal. Very
      ;; useful for long, single lines.
      (setq line-move-visual t)

      ;; differentiate buffers with identical names
      (use-package uniquify
        :ensure nil
        :config
        (setq uniquify-buffer-name-style 'post-forward-angle-brackets))


      ;; use newer files rather than old compiled files, .elc
      (setq load-prefer-newer t)

      ;; if emacs crashes w/o warning, we want a valid list of recent files, don't we?
      (run-at-time nil (* 5 60) 'recentf-save-list)

      ;; finally, set emacs to display the path directory in the menu bar
      ;; displays current working directory at all times in emacs
      (setq frame-title-format
            (list (format "%s %%S: %%j " (system-name))
                  '(buffer-file-name "%f" (dired-directory dired-directory "%b"))))
    #+END_SRC
**** garbage collection & read process throughput
     emacs' default gc threshold is set an extraordinarily small value, like 80KB, which is great
     in terms of performance relative to other text editors sans Vim; however, since we use [[lsp
     mode]], to get performance on par with other IDE's like VSCode, we need to increase it a fair
     amount. This is per the [[https://emacs-lsp.github.io/lsp-mode/page/performance/][LSP recommendations themselves]]:
     #+BEGIN_SRC emacs-lisp
     (setq gc-cons-threshold 150000000) ;; ~150mb
     #+end_src

     for a similar reason, we need to increase the amount of data that emacs can read from a single
     process. The default is at 4KB when we need somewhere between 800KB-3MB:
     #+BEGIN_SRC emacs-lisp
     (setq read-process-output-max (* 1024 1024)) ;; 1mb
     #+end_src
**** truncate line fringe
     by default, truncate lines will show a little arrow pointing right when a line is longer than
     the visual line of the current frame. when i use fill-column-indicater mode (fci), which i do
     for basically all programming buffers, this results in the entire right fringe being covered in
     arrows whenever the frame is of a smaller width than the fill column i've set for that given
     language. this tweak here simply removes the fringe indicater for line truncation because,
     afaik, i've never needed emacs to warn me of a line truncation, at least when using the GUI
     version.
     #+begin_src emacs-lisp
       (setq-default fringe-indicator-alist (assq-delete-all 'truncation fringe-indicator-alist))
     #+end_src
     the above tweak was taken from [[https://www.reddit.com/r/emacs/comments/2t16il/is_there_a_way_to_remove_the_truncated_lines/cnurpo4?utm_source=share&utm_medium=web2x][sergeinosov]]
**** delete selection mode
     if you C-a while in a text box on a browser like firefox and then type anything, whether the
     character "f" or hit the backspace key, the whole selection of text is deleted. emacs, by
     default, doesn't do this. instead, when a region is marked, it just appends the text and
     deactivates the selection. in newish versions of emacs, though, you can enable the behavior you
     typically expect via =delect-section-mode=. that's what we're doing here.

     #+begin_src emacs-lisp
       (delete-selection-mode 1)
     #+end_src
**** limit buffer name size in modeline
     Sometimes I have very long file names (think textbook pdfs) and it makes
     seeing anything else on my mode line impossible. So let's reduce what can
     be shown:
     #+BEGIN_SRC emacs-lisp
       (setq-default mode-line-buffer-identification (list -40 (propertized-buffer-identification "%12b")))
     #+END_SRC
**** which-key and general.el
     Which-key is a very nice minor-mode that essentially gives a preview of
     all possible key completions available based on the current key stroke,
     saving you the hassle of having to memorize *every* key binding, and thus
     protecting you from going full emacs hermit.


     general defines "leader" keys/keystrokes, which allows you to setup mappings
     predicated off of a set of leader keys. For example, I set "C-c m" to be
     the leader keystroke for all of my magit commands I typically use.

     key bindings are left to the actual key binding section

     #+BEGIN_SRC emacs-lisp
       (use-package general)

       (use-package which-key
         :diminish ""
         :init
         (which-key-mode)
         :config
         ;; already by default, but explicitly setting for popup to be in a bottom
         ;; frame.
         (which-key-setup-side-window-bottom)
         ;; ensures popup cannot take up more than .15 of emacs window
         (setq which-key-side-window-max-height .10))
     #+END_SRC
**** fill-column, spaces, tabs, and unfilling paragraphs
     I like my line wrapping, generally speaking, at around 80 characters. I change this for
     different languages, but when in markup/markdown langs like org-mode I still prefer my lines
     wrapping at around 80 chars.

     unfill helps with making a set of lines that I select be formed into a proper paragraph that
     wraps at the current fill-column value.
     #+BEGIN_SRC emacs-lisp
       (setq-default fill-column 79)
       (use-package unfill)
     #+END_SRC


     I generally don't use tabs and what them to be treated with a width of 2 when they do appear.
     #+begin_src emacs-lisp
       (setq-default indent-tabs-mode nil)
       (setq-default default-tab-width 2)
     #+end_src
**** theme, styling, and transparency
     #+BEGIN_SRC emacs-lisp
       ;; Set transparency of emacs
       (defun transparency (value)
         "Sets the transparency of the frame window. 0=transparent/100=opaque"
         (interactive "nTransparency Value 0 - 100 opaque:")
         (set-frame-parameter (selected-frame) 'alpha value))

       ;; i have one them i play with that isn't on melpa
       (add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")

       (use-package colorless-themes
         :quelpa (colorless-themes :fetcher github :repo "lthms/colorless-themes.el"))

       (use-package kaolin-themes)
       (use-package tao-theme)

       (load-theme 'minimal-light t)


       ;; required to have custom-theme-set-faces work again: https://emacs.stackexchange.com/a/52804
       (setq custom--inhibit-theme-enable nil)

       (custom-theme-set-faces
        'minimal-light
        '(region ((t (:foreground "gainsboro" :background "dim grey"))))
        '(show-paren-match ((t (:background "gray30" :foreground "white" :weight extrabold))))
        '(org-block-begin-line ((t (:foreground "gray70" :background "gray98"))))
        '(org-block-end-line ((t (:foreground "gray70" :background "gray98"))))
        '(tooltip ((t (:background "gray90" :height 0.95 :family "Fira Code"))))
        '(italic ((t (:font "Hack" :slant italic :height 1.05))))
        '(lsp-ui-sideline-code-action ((t (:foreground "DarkRed"))))
        '(lsp-ui-sideline-current-symbol ((t (:foreground "DarkBlue"))))
        '(lsp-ui-sideline-symbol ((t (:foreground "Black"))))
        '(lsp-ui-doc-background ((t (:background "gray92"))))
        '(aw-leading-char-face ((t (:background "white smoke" :foreground "red" :box (:line-width 1 :color "black") :height 2.5))))
        '(table-cell ((t (:background "LightSteelBlue" :foreground "black"))))
        '(fringe ((t (:background "gray100"))))
        '(lsp-ui-doc-border ((t (:border "black"))))
        )

       ;; pretty symbols
       (setq-default prettify-symbols-alist '(("lambda" . ?λ)
                                              (".\\" . ?λ)
                                              ("->" . ?⟶)
                                              ("=/=" . ?≠)
                                              ("===" . ?≡)
                                              ("!=" . ?≠)
                                              ("<=" . ?≤)
                                              (">=" . ?≥)
                                              ("alpha" . ?α)
                                              ("beta" . ?β)
                                              ("Gamma" . ?Γ)
                                              ("gamma" . ?γ)
                                              ("Delta" . ?Δ)
                                              ("delta" . ?δ)
                                              ("epsilon" . ?ε)
                                              ("theta" . ?θ)
                                              ("mu" . ?μ)
                                              ("Xi" . ?Ξ)
                                              ("xi" . ?ξ)
                                              ("Pi" . ?Π)
                                              ("pi" . ?π)
                                              ("rho" . ?ρ)
                                              ("Sigma" . ?Σ)
                                              ("sigma" . ?σ)
                                              ("tau" . ?τ)
                                              ("Phi" . ?Φ)
                                              ("phi" . ?φ)
                                              ("psi" . ?ψ)
                                              ("Omega" . ?Ω)
                                              ("omega" . ?ω)
                                              ;; ("x" . x)
                                              ))
       ;; dims the buffers that are not active

.emacs.d/config.el

(let* ((package--builtins nil)
       (packages
        '(
          ;; utilities
          general which-key ivy swiper counsel fill-column-indicator volatile-highlights
          golden-ratio undo-tree ace-window paredit projectile diminish rainbow-delimiters
          smartparens ggtags expand-region ;;bookmarks+ dired+ no longer on melpa 


          ;; auto-complete
          company flycheck flycheck-tip hippie-exp yasnippet auto-yasnippet flycheck-irony company-c-headers
          elisp-slime-nav company-math company-auctex company-anaconda company-tern
          slime-company company-ycmd flycheck-rust enh-ruby-mode
          company-ghci flycheck-haskell company-statistics

          ;; git
          git-gutter-fringe magit

          ;; markdown
          markdown-mode toml-mode yaml-mode

          ;; c/cpp
          clang-format cmake-mode

          ;; rust
          cargo racer rust-mode

          ;; lisp
          slime racket-mode

          ;; clojure
          cider cider-eval-sexp-fu clj-refactor clojure-mode clojure-snippets

          ;; haskell
          haskell-mode haskell-snippets hindent hlint-refactor

          ;; python
          anaconda-mode pyenv-mode hy-mode pytest

          ;; js
          js-doc js2-mode js2-refactor json-mode json-snatcher web-beautify coffee-mode

          ;; webprogramming
          haml-mode sass-mode scss-mode slim-mode web-mode less-css-mode pug-mode emmet-mode


          ;; ruby
          bundler chruby rbenv rspec-mode rubocop ruby-tools rvm rake

          ;; tex
          auctex auctex-latexmk typo

          ;; org
          org-bullets org-pomodoro org-download htmlize org-journal

          ;; themes
          seti-theme gandalf-theme cyberpunk-theme dark-mint-theme dracula-theme

          ;; other
          writeroom-mode minimap gnuplot csv-mode mips-mode smart-mode-line fireplace

          ;; google
          google-this google-translate

          )))
  (let ((packages (remove-if 'package-installed-p packages)))
    (when packages
      ;; Install uninstalled packages
      (package-refresh-contents)
      (mapc 'package-install packages))))

;; no splash screen, sorry Stallman
(setq inhibit-splash-screen t)

;; don't use the tool or scroll bar, thx
(tool-bar-mode -1)
(scroll-bar-mode -1)

;; do like menu bar
(menu-bar-mode 1)

;; UTF-8 is our friend in a world of shitty programming standards
(prefer-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
(setq default-buffer-file-coding-system 'utf-8)

;; speaking of text, unify emacs clipboard with linux
(setq x-select-enable-clipboard t)

;; global hl line mode
(global-hl-line-mode t)

;; can't get flycheck mode to enable itself otherwise
(global-flycheck-mode)

;; make our mode line prettier
(use-package smart-mode-line)
(setq sml/no-confirm-load-theme t)
(sml/setup)

;; make sure it attempts utf-8 first when pasting text into emacs
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

;; always tell me the column and row of where i am please, i am already lost enough as is.
(line-number-mode 1)
(column-number-mode 1)

;; do you seriously think i'm good at typing
(setq read-file-name-completion-ignore-case t)

;; life is too short for fully authenticating bad decisions
(defalias 'yes-or-no-p 'y-or-n-p)

;; fix bad escape sequence weirdness
(setq system-uses-terminfo nil)

;; stronk encryption good
(setq gnutls-min-prime-bits 4096)

;; don't let me be moronic and kill emacs w/o warning. GUI only.
(when (window-system)
  (setq confirm-kill-emacs 'yes-or-no-p))

;; move through lines based on visual appearance rather than literal. Very useful for long, single lines.
(setq line-move-visual t)

;; differentiate buffers with identical names
(use-package uniquify
  :config
  (setq uniquify-buffer-name-style 'post-forward-angle-brackets))

;; use newer files rather than old compiled files, .elc
(setq load-prefer-newer t)

;; if emacs crashes w/o warning, we want a valid list of recent files, don't we?
(run-at-time nil (* 5 60) 'recentf-save-list)

;; finally, set emacs to display the path directory in the menu bar
;; displays current working directory at all times in emacs
(setq frame-title-format
      (list (format "%s %%S: %%j " (system-name))
            '(buffer-file-name "%f" (dired-directory dired-directory "%b"))))

;; lisp and haskell got me used to working with 2 space indents, idk man
(setq-default fill-column 80)
(setq-default indent-tabs-mode nil)
(setq-default default-tab-width 2)
(add-hook 'prog-mode-hook 'linum-mode)

;; Set transparency of emacs
(defun transparency (value)
  "Sets the transparency of the frame window. 0=transparent/100=opaque"
  (interactive "nTransparency Value 0 - 100 opaque:")
  (set-frame-parameter (selected-frame) 'alpha value))

;; for dark: seti; for light: gandalf
;; when i feel like a hacker: cyberpunk or dark-mint
(load-theme 'dracula t)

;; pretty symboles
(prettify-symbols-mode t)
(setq-default prettify-symbols-alist '(("lambda" . ?λ)
                                       ("delta" . ?Δ)
                                       ("gamma" . ?Γ)
                                       ("phi" . ?φ)
                                       ("psi" . ?ψ)))

;; C-c LEFT to undo window change, which i need often
(use-package winner
  :init (winner-mode 1))

;; re-opens file at last place edited
(use-package saveplace
  :defer t
  :init
  (setq-default save-place t)
  (setq save-place-file (expand-file-name ".places" user-emacs-directory)))

;; enables recent-files to be re-opened
(recentf-mode 1)
(setq recentf-max-menu-items 50)

;; do not need spam in recents list
(setq recentf-exclude '("/auto-install/" ".recentf" "/repos/" "/elpa/"
                        "\\.mime-example" "\\.ido.last" "COMMIT_EDITMSG"
                        ".gz"
                        "~$" "/tmp/" "/ssh:" "/sudo:" "/scp:"))
(global-set-key "\C-x\ \C-r" 'recentf-open-files)

(ivy-mode 1)
(diminish 'ivy-mode)
(setq ivy-use-virtual-buffers t)
(setq enable-recursive-minibuffers t)
(global-set-key "\C-s" 'swiper)
(global-set-key (kbd "C-c C-r") 'ivy-resume)
(global-set-key (kbd "<f6>") 'ivy-resume)
(global-set-key (kbd "M-x") 'counsel-M-x)
(global-set-key (kbd "C-x C-f") 'counsel-find-file)
(global-set-key (kbd "<f1> f") 'counsel-describe-function)
(global-set-key (kbd "<f1> v") 'counsel-describe-variable)
(global-set-key (kbd "<f1> l") 'counsel-find-library)
(global-set-key (kbd "<f2> i") 'counsel-info-lookup-symbol)
(global-set-key (kbd "<f2> u") 'counsel-unicode-char)
(global-set-key (kbd "C-c g") 'counsel-git)
(global-set-key (kbd "C-c j") 'counsel-git-grep)
(global-set-key (kbd "C-c k") 'counsel-ag)
(global-set-key (kbd "C-x l") 'counsel-locate)
(global-set-key (kbd "C-S-o") 'counsel-rhythmbox)
(define-key minibuffer-local-map (kbd "C-r") 'counsel-minibuffer-history)

(defun my/dired-mode-hook ()
  (toggle-truncate-lines 1))

(use-package dired
  :bind ("C-x C-j" . dired-jump)
  :config
  (progn
    (use-package dired-x
      :init (setq-default dired-omit-files-p t)
      :config
      (add-to-list 'dired-omit-extensions ".DS_Store"))
    (customize-set-variable 'diredp-hide-details-initially-flag nil)
    ;; (use-package dired+) no longer available via melpa
    (use-package dired-aux
      :init (use-package dired-async))
    (put 'dired-find-alternate-file 'disabled nil)
    (setq ls-lisp-dirs-first t
          dired-recursive-copies 'always
          dired-recursive-deletes 'always
          dired-dwim-target t
          ;; -F marks links with @
          dired-ls-F-marks-symlinks t
          delete-by-moving-to-trash t
          ;; Auto refresh dired
          global-auto-revert-non-file-buffers t
          wdired-allow-to-change-permissions t)
    (add-hook 'dired-mode-hook #'my/dired-mode-hook)))

(use-package flyspell
  :init
  (add-hook 'text-mode-hook 'turn-on-flyspell)
  (add-hook 'prog-mode-hook 'flyspell-prog-mode))

  ;; make ispell fast and make it only look at 3 char+ words
  (setq ispell-extra-args
        (list "--sug-mode=fast" ;; ultra|fast|normal|bad-spellers
              "--lang=en_US"
              "--ignore=3"))

;; whitespace checker starts at 80
(setq whitespace-line-column 80)

;; what whitespace looks for
(setq whitespace-style '(tabs newline space-mark
                   tab-mark newline-mark
                   face lines-tail))

;; special visual market up for non-whitespace
(setq whitespace-display-mappings
      ;; all numbers are Unicode codepoint in decimal. e.g. (insert-char 182 1)
      ;; 32 SPACE, 183 MIDDLE DOT
      '((space-mark nil)
        ;; 10 LINE FEED
        ;;(newline-mark 10 [172 10])
        (newline-mark nil)
        ;; 9 TAB, MIDDLE DOT
        (tab-mark 9 [183 9] [92 9])))

;; disabled for modes that it doesn't make sense
(setq whitespace-global-modes '(not org-mode
                                   eshell-mode
                                   shell-mode
                                   web-mode
                                   log4j-mode
                                   "Web"
                                   dired-mode
                                   emacs-lisp-mode
                                   clojure-mode
                                   lisp-mode))

(use-package company
  :defer t
  :diminish ""
  :bind ("C-." . company-complete)
  :init (add-hook 'prog-mode-hook 'company-mode)
  :config
  (progn
    (setq company-idle-delay 0.1
          company-minimum-prefix-length 2
          company-selection-wrap-around t
          company-dabbrev-downcase nil
          company-transformers '(company-sort-by-occurrence))
    (bind-keys :map company-active-map
               ("C-n" . company-select-next)
               ("C-p" . company-select-previous)
               ("C-d" . company-show-doc-buffer)
               ("<tab>" . company-complete)))
  (add-hook 'after-init-hook 'company-statistics-mode))

;; haskell
(eval-after-load "company"
  '(add-to-list 'company-backends 'company-ghci))

;; python
(eval-after-load "company"
  '(add-to-list 'company-backends '(company-anaconda :with company-capf)))

(use-package expand-region
  :bind ("M-/" . er/expand-region))

(defun my-flycheck-customize ()
  (interactive)
  (global-set-key (kbd "C-c C-n") 'flycheck-tip-cycle)
  (global-set-key (kbd "C-c C-p") 'flycheck-tip-cycle-reverse))

(use-package flycheck
  :defer t
  :bind (("M-g M-n" . flycheck-next-error)
         ("M-g M-p" . flycheck-previous-error)
         ("M-g M-=" . flycheck-list-errors))
  :diminish ""
  :config
  (use-package flycheck-tip
    :config (add-hook 'flycheck-mode-hook 'my-flycheck-customize)))

(use-package undo-tree
  :init (global-undo-tree-mode t)
  :defer t
  :diminish ""
  :config
  (progn
    (define-key undo-tree-map (kbd "C-x u") 'undo-tree-visualize)
    (define-key undo-tree-map (kbd "C-/") 'undo-tree-undo)
    (define-key undo-tree-map (kbd "C-M-/") 'undo-tree-redo)))

(defun generic-org-minor-modes ()
  (interactive)
  (org-bullets-mode 1))
(setq org-journal-dir "/home/spook/Documents/.journal/")
(add-hook 'org-mode-hook 'generic-org-minor-modes)

(defun my-add-watchwords ()
  "Highlight FIXME, TODO, and NOCOMMIT in code"
  (font-lock-add-keywords
   nil '(("\\<\\(FIXME\\|TODO\\|NOCOMMIT\\)\\>"
          1 '((:foreground "#d7a3ad") (:weight bold)) t))))

(add-hook 'prog-mode-hook 'my-add-watchwords)

(use-package haskell-mode
  :defer t
  :config
  (defun my-haskell-setup()
    (interactive)
    (haskell-doc-mode)
    (haskell-indent-mode)
    (flycheck-haskell-setup)
    ;;(hindent-mode) ;; must install with stack
    (haskell-snippets-initialize)
    (setq haskell-process-type 'stack-ghci))
  (add-hook 'haskell-mode-hook 'my-haskell-setup))

(use-package python
  :defer t
  :config
  (defun my-python-setup ()
    (interactive)
    (anaconda-mode)
    (anaconda-eldoc-mode)
    (pyenv-mode)
    ;; (pytest-mode)
    ;; (semantic-mode 0)
    (yapf-mode))
  (add-hook 'python-mode-hook 'my-python-setup))

(defun activate-slime-helper ()
  (when (file-exists-p "~/.quicklisp/slime-helper.el")
    (load (expand-file-name "~/.quicklisp/slime-helper.el"))
    (define-key slime-repl-mode-map (kbd "C-l")
      'slime-repl-clear-buffer))
  (remove-hook 'lisp-mode-hook #'activate-slime-helper))

(add-hook 'lisp-mode-hook #'activate-slime-helper)

(setq inferior-lisp-program "sbcl")

(setq lisp-loop-forms-indentation   6
      lisp-simple-loop-indentation  2
      lisp-loop-keyword-indentation 6)

(defun my/helpful-lisp-modes ()
  (interactive)
  (paredit-mode 1)
  (rainbow-delimiters-mode 2)
  (show-paren-mode 1)
  (eldoc-mode 1))

(add-hook 'lisp-mode-hook #'my/helpful-lisp-modes)

(defun my/turn-on-paredit-and-eldoc ()
  (interactive)
  (paredit-mode 1)
  (rainbow-delimiters-mode 1)
  (show-paren-mode 1)
  (eldoc-mode 1))

(add-hook 'emacs-lisp-mode-hook #'my/turn-on-paredit-and-eldoc)
(add-hook 'ielm-mode-hook #'my/turn-on-paredit-and-eldoc)

(use-package eldoc
  :config
  (progn
    (use-package diminish
      :init
      (progn (diminish 'eldoc-mode "")))
    (setq eldoc-idle-delay 0.3)
    (set-face-attribute 'eldoc-highlight-function-argument nil
                        :underline t :foreground "green"
                        :weight 'bold)))

(defun ielm-other-window ()
  "Run ielm on other window"
  (interactive)
  (switch-to-buffer-other-window
   (get-buffer-create "*ielm*"))
  (call-interactively 'ielm))

(define-key emacs-lisp-mode-map (kbd "C-c C-z") 'ielm-other-window)
(define-key lisp-interaction-mode-map (kbd "C-c C-z") 'ielm-other-window)

(bind-key "M-:" 'pp-eval-expression)

(defun sanityinc/eval-last-sexp-or-region (prefix)
  "Eval region from BEG to END if active, otherwise the last sexp."
  (interactive "P")
  (if (and (mark) (use-region-p))
      (eval-region (min (point) (mark)) (max (point) (mark)))
    (pp-eval-last-sexp prefix)))

(bind-key "C-x C-e" 'sanityinc/eval-last-sexp-or-region emacs-lisp-mode-map)

(define-key lisp-mode-shared-map (kbd "RET") 'reindent-then-newline-and-indent)

(use-package rust-mode
  :defer t
  :config
  (defun my-rust-setup ()
    (interactive)
    (racer-mode)
    (cargo-minor-mode)
    (flycheck-rust-setup))
  (add-hook 'rust-mode-hook 'my-rust-setup))

(use-package web-mode
  :defer t
  :config
  (defun my-web-mode-setup ()
    (interactive)
    (emmet-mode)
    (compan)
    ()))

(add-to-list 'auto-mode-alist '("\\.tex\\'" . latex-mode))
(add-to-list 'auto-mode-alist '("\\.tex\\'" . latex-mode))
(use-package tex
  :ensure auctex)


